matrix_sdk_base/src/client.rs

```diff
@@ -1320,6 +1373,130 @@ impl BaseClient {
+            .await
+            .unwrap()
+        {
+            event.content.global
+        } else {
+            // FIXME don't panic if the user is not logged in?
+            let session = self.get_session().await.unwrap();
+            Ruleset::server_default(&session.user_id)
+        }
+    }
+
+    /// Get the push context for the given room.
+    ///
+    /// Tries to get the data from `changes` or the up to date `room_info`. Loads the data from the
+    /// store otherwise.
+    ///
+    /// Returns `None` if some data couldn't be found. This should only happen in brand new rooms,
+    /// while we process its state.
+    pub async fn get_push_room_context(
+        &self,
+        room: &Room,
+        room_info: &RoomInfo,
+        changes: &StateChanges,
+    ) -> Result<Option<PushConditionRoomCtx>> {
+        let room_id = room.room_id();
+        let user_id = room.own_user_id();
+
+        let member_count = room_info.active_members_count();
+
+        let user_display_name = if let Some(member) = changes
+            .members
+            .get(room_id)
+            .and_then(|members| members.get(user_id))
+        {
+            let member = member.clone();
+            member
+                .content
+                .displayname
+                .unwrap_or_else(|| user_id.localpart().to_owned())
+        } else if let Some(member) = room.get_member(user_id).await? {
+            member.name().to_owned()
+        } else {
+            return Ok(None);
+        };
+
+        let room_power_levels = if let Some(AnySyncStateEvent::RoomPowerLevels(event)) = changes
+            .state
+            .get(room_id)
+            .and_then(|types| types.get(&EventType::RoomPowerLevels.to_string()))
+            .and_then(|events| events.get(""))
+        {
+            event.content.clone()
+        } else if let Some(AnySyncStateEvent::RoomPowerLevels(event)) = self
+            .store
+            .get_state_event(room_id, EventType::RoomPowerLevels, "")
+            .await?
+        {
+            event.content
+        } else {
+            return Ok(None);
+        };
+
+        Ok(Some(PushConditionRoomCtx {
+            room_id: room_id.clone(),
+            member_count: UInt::new(member_count).unwrap(),
+            user_display_name,
+            users_power_levels: room_power_levels.users,
+            default_power_level: room_power_levels.users_default,
+            notification_power_levels: room_power_levels.notifications,
+        }))
+    }
+
+    /// Update the push context for the given room.
+    ///
+    /// Updates the context data from `changes` or `room_info`.
+    pub async fn update_push_room_context(
+        &self,
+        push_rules: &mut PushConditionRoomCtx,
+        user_id: &UserId,
+        room_info: &RoomInfo,
+        changes: &StateChanges,
+    ) {
+        let room_id = &room_info.room_id;
+
+        push_rules.member_count = UInt::new(room_info.active_members_count()).unwrap();
```

Comment:
The unwrap needs to be gone, not sure what makes sense here, perhaps clamping is ok? Probably depends on how the member count is used.

Reply:
It is used to compare with a `PushCondition::RoomMemberCount`, so the comparison is with another `UInt`. In the default push rules it's used to check if the room is "one to one".

So I guess clamping is OK. Is there an easy way to do this, since I don't see a method for that on `UInt`?

Reply:
`UInt::new(u64).unwrap_or(UInt::MAX)`